Sources/Plasma/Apps/plClient/win32/plWinDpi.cpp

```diff
@@ -0,0 +1,287 @@
+#include "pnMessage/plClientMsg.h"
+
+#ifndef WM_GETDPISCALEDSIZE
+#   define WM_GETDPISCALEDSIZE 0x02E4
+#endif
+
+static std::unique_ptr<plWinDpi> s_instance;
+
+plWinDpi& plWinDpi::Instance()
+{
+    if (!s_instance)
+        s_instance = std::make_unique<plWinDpi>();
+    return *s_instance.get();
+}
+
+plWinDpi::plWinDpi()
+    : fAdjustWindowRectExForDpi(L"user32", "AdjustWindowRectExForDpi"),
+      fAreDpiAwarenessContextsEqual(L"user32", "AreDpiAwarenessContextsEqual"),
+      fEnableNonClientDpiScaling(L"user32", "EnableNonClientDpiScaling"),
+      fGetDpiForMonitor(L"shcore", "GetDpiForMonitor"),
+      fGetDpiForSystem(L"user32", "GetDpiForSystem"),
+      fGetDpiForWindow(L"user32", "GetDpiForWindow"),
+      fGetSystemDpiForProcess(L"user32", "GetSystemDpiForProcess"),
+      fGetSystemMetricsForDpi(L"user32", "GetSystemMetricsForDpi"),
+      fGetThreadDpiAwarenessContext(L"user32", "GetThreadDpiAwarenessContext"),
+      fSetProcessDpiAwareness(L"shcore", "SetProcessDpiAwareness"),
+      fSetProcessDpiAwarenessContext(L"user32", "SetProcessDpiAwarenessContext")
+{
+    Initialize();
+}
+
+void plWinDpi::Initialize() const
+{
+    // There are three different levels of DPI awareness that we can deal with,
+    // each was added in newer versions of Windows as monitors became more and
+    // more sexy.
+    // In Vista, the entire system had an adjustable display scale.
+    // In Windows 8, each monitor was allowed to have a separate scale.
+    // In Windows 10, the per-monitor scaling was slowly improved to automatically
+    // scale dialog boxes and things like that.
+    // So we need to try to handle all of that. We will assume the baseline is Vista.
+    // Any API added later than that will need to be called on a provisional basis.
+    LogWhite("--- Begin DPI Awareness ---");
+    do {
+        // Per Monitor V2 for Windows 10 v1703.
+        auto perMonitorV2Result = fSetProcessDpiAwarenessContext(DPI_AWARENESS_CONTEXT_PER_MONITOR_AWARE_V2);
+        if (perMonitorV2Result.has_value()) {
+            if (perMonitorV2Result.value() == TRUE) {
+                LogGreen("Using Per Monitor v2");
+                break;
+            }
+            LogRed("Failed to set DPI Awareness Per Monitor v2: {}", hsCOMError(hsLastWin32Error, GetLastError()));
+        } else {
+            LogYellow("Per Monitor v2 isn't supported on this OS.");
+        }
+
+        // Per Monitor v1 for Windows 8.1.
+        auto perMonitorV1Result = fSetProcessDpiAwareness(PROCESS_PER_MONITOR_DPI_AWARE);
+        if (perMonitorV1Result.has_value()) {
+            if (SUCCEEDED(perMonitorV1Result.value())) {
+                LogGreen("Using Per Monitor V1");
+                break;
+            } else {
+                LogRed("Failed to set DPI Awareness Per Monitor v1: {}", hsCOMError(perMonitorV1Result.value()));
+            }
+        }
+
+        // System DPI Awareness.
+        if (SetProcessDPIAware() != FALSE) {
```

Comment:
Hey @Hoikas - a few questions...

I noticed this function wasn't wrapped in an optional win call - could this cause trouble on Windows XP in since this function is only available on Vista?

I'm also looking at implementing the minimum functionality needed in the patcher to set the DPI of the app. Don't have a problem implementing it - just curious if there is a reason to query/go to the newer API and if that's relevant to the patcher.

Reply:
Our dependencies aren't compatible with Windows XP anymore, so I don't think that's much of a concern at this point.

Reply:
Ah - is OpenUru also dropping Windows XP, or is that a larger conversation?

Reply:
I think it's currently a non-conversation because you know how talking to them goes... Realistically, we're already having to patch Python 3.10 to retain Windows 7 compatibility so the ship has to sail at some point. If we ever update to PhysX 5.1, they've dropped 32-bit support.

Reply:
I think @dpogue has already answered everything here, but, if you want, you can change this to use the optional wincall thing to maintain compatibility, but that's not really something that we really care about in general.

I would really prefer to see the code shared as much as possible - do you think you could reorganize `plWinDpi` to not depend on `plMessage` instead of duplicating the code?

Reply:
Let me think it through. I could probably move the core parts into their own module while making sure the stuff that is only needed for the engine lands in the engine. Maybe I could put a base class in it's own module and then extend the Plasma specific class with the rest.